src/callbacks_stage/positivity_adaptive_filter_dg1d.jl

```diff
@@ -0,0 +1,171 @@
+
+  try
+    calc_local_filtered_values!(cache, view(u, :, e), θ, e, filter, mesh, equations, dg)
+    return check_local_bound(e, local_bound, equations, dg, cache) 
+  catch err
+    # Catch negativity during entropy projection, else throw other errors
+    isa(err, DomainError) ? false : throw(err)
+  end
+
+end
+
+function calc_local_filtered_values!(cache, u_e, θ, e, filter, mesh, equations, dg)
+
+  calc_local_filtered_cons_values!(cache, u_e, θ, e, filter, mesh, equations, dg)
+  local_entropy_projection!(cache, u_e, e, mesh, equations, dg)
+
+end
+
+function calc_local_filtered_cons_values!(cache, u_e, θ, e, filter, mesh, equations, dg)
+
+  rd = dg.basis
+  @unpack VDM = rd
+  @unpack local_u_modal_coeffs_threaded = cache
+
+  apply_local_filter!(cache, θ, e, filter, mesh, dg)   # Update filter value into u_modal_coeffs_threaded
+  apply_to_each_field(mul_by!(VDM), u_e, local_u_modal_coeffs_threaded[Threads.threadid()])
+
+end
+
+# TODO: Only support 1D. Dispatch on dimension in future PRs
+function apply_local_filter!(cache, θ, e, filter::SecondOrderExponentialAdaptiveFilter, mesh::DGMultiMesh{1}, dg::DGMulti{1})
+  
+  @unpack u_modal_coeffs, local_u_modal_coeffs_threaded = cache
+
+  for i in each_mode(mesh, dg)
+    local_u_modal_coeffs_threaded[Threads.threadid()][i] = exp(-θ*(i-1)*(i-1))*u_modal_coeffs[i, e]
+  end
+
+end
+
+function apply_local_filter!(cache, θ, e, filter::ZhangShuScalingAdaptiveFilter, mesh::DGMultiMesh{1}, dg::DGMulti{1})
+
+  @unpack u_modal_coeffs, local_u_modal_coeffs_threaded = cache
+
+  for i in each_mode(mesh, dg)
+    i >= 2 ? local_u_modal_coeffs_threaded[Threads.threadid()][i] = θ*u_modal_coeffs[i, e] : nothing
+  end
+
+end
+
+function check_local_bound(e, local_bound, equations, dg, cache)
+
+  @unpack entropy_projected_u_values, entropy_var_values = cache
+
+  rd = dg.basis
+  entropy_var_face_values_e    = view(entropy_var_values, :, e)
+  entropy_projected_u_values_e = view(entropy_projected_u_values, :, e)
+
+  rho     = Base.Generator(u->density(u, equations), entropy_projected_u_values_e)
+  rhoe    = Base.Generator(u->energy_internal(u, equations), entropy_projected_u_values_e)
+  vf_last = Base.Generator(v->v[end], entropy_var_face_values_e)
+
+  return (mapreduce((x -> x >= local_bound.ρmin)    , &, rho    )
+       && mapreduce((x -> x <= local_bound.ρmax)    , &, rho    )
+       && mapreduce((x -> x >= local_bound.ρemin)   , &, rhoe   )
+       && mapreduce((x -> x <= local_bound.ρemax)   , &, rhoe   )
+       && mapreduce((x -> x <= local_bound.vlastmax), &, vf_last))
+
+end
+
+# TODO: REFACTOR. Not so obvious on how to use Roots.jl in this case 
+#####################
+###   Utilities   ###
+#####################
+
+"""
+Bisection algorithm to find θ closest to θ_invalid that cond(θ) == true.
+Assuming cond(θ_valid) == true, and cond is a continuous function in the
+interval between θ_valid and θ_invalid
+"""
+function bisection_bound(cond, θ_valid, θ_invalid)
```

Comment:
I didn't find a way to use Roots.jl in this case, can you give me a hint?

Reply:
I just meant that instead of calling bisection_bound you could use a rootfinder from Roots.jl?

Reply:
"Roots is a Julia package for finding zeros of continuous scalar functions of a single real variable.". The function `cond` in our case is not a scalar?

Reply:
Ah, I see - I was imagining that you would use a scalar rootfinder for each bound specified and take the minimum of the filtering parameters. Does the vector approach reduce costs compared with the scalar approach?

Reply:
Yes - I cannot precalculate the entropy-projected variables before the nonlinear solve, since nonlinear solves change filtering factors. I need to do a new entropy projection in each iteration in order to determine, for example, entropy-projected variable. But another problem is the negativity - imagine we want to bound entropy projected internal energy `rhoe(u_tilde) >= 0`, and we have two filter factors `theta = 0` (reduce to high order scheme) and `theta = Inf` (reduce to low order scheme). I feed it in Roots.jl hoping to find `theta` such that `rhoe(u_tilde(theta)) = 0`, what will happen when I try to evaluate `rhoe(u_tilde(0))` if high order scheme has negative values?

Reply:
Wouldn't theta be between `[0, 1]` using the adaptive filtering trick?

I would guess that you'd get a `DomainError` related to the negative internal energy, but I think this would be easier to discuss in person...